In [4]:
from skimage import io
from skimage.transform import resize
from skimage.color import rgb2grey
from os import listdir
import numpy as np

Here we load in our presorted data as images labeled good or bad.

In [5]:
goodFiles = listdir("GoodArt")
badFiles = listdir("BadArt")

Below we load in each of our images, convert the image to gray scale, resize to a 256 by 256 pixel, and flatten into row vector. We keep a counter to keep track of how many images were processed.

In [6]:
goodArt_Transformed = []
i = 1
for file in goodFiles:
    print(i)
    #take each image, convert to gray scale, resize all images to same size
    #flatten so that on row is one image
    if ".jpg" in file:
        i = i +1
        file = "GoodArt/"+file
        original = io.imread(file)
        original = rgb2grey(original)
        good_tmp = resize(original,(256,256))
        good_tmp = good_tmp.flatten()
        goodArt_Transformed.append(good_tmp)

1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
94
95
96
97
98
99
100
101
102
103
104
105
106
107
108
109
110
111
112
113
114
115
116
117
118
119
120
121
122
123
124
125
126
127
128
129
130
131
132
133
134
135
136
137
138
139
140
141
142
143
144
145
146
147
148
149
150
151
152
153
154
154
155
156
157
158
159
160
161
162
163
164
165
166
167
168
169
170
171
172
173
174
175
176
177
178
179
180
181
182
183
184
185
186
187
188
189
190
191
192
193
194
195
196
197
198
199
200
201
202
203
204
205
206
207
208
209
210
211
212
213
214
215
216
217
218
219
220
221
222
223
224
225
226
227
228
229
230
231
232
233
234
235
236
237
238
239
240
241
242
243
244
245
246
247
248
249
250
251
252
253
254
255
256
257
258
259
260
261
262
263
264
265
266
267
268
269
270
271
272
273
274
275
2

In [7]:
badArt_Transformed = []
i = 1
for file in badFiles:
    if ".jpg" in file:
        print(i)
        i = i +1
        file = "BadArt/"+file
        original = io.imread(file)
        original = rgb2grey(original)
        bad_tmp = resize(original,(256,256))
        bad_tmp = bad_tmp.flatten()
        badArt_Transformed.append(bad_tmp)

1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99
100
101
102
103
104
105
106
107
108
109
110
111
112
113
114
115
116
117
118
119
120
121
122
123
124
125
126
127
128
129
130
131
132
133
134
135
136
137
138
139
140
141
142
143
144
145
146
147
148
149
150
151
152
153
154
155
156
157
158
159
160
161
162
163
164
165
166
167
168
169
170
171
172
173
174
175
176
177
178
179
180
181
182
183
184
185
186
187
188
189
190
191
192
193
194
195
196
197
198
199
200
201
202
203
204
205
206
207
208
209
210
211
212
213
214
215
216
217
218
219
220
221
222
223
224
225
226
227
228
229
230
231
232
233
234
235
236
237
238
239
240
241
242
243
244
245
246
247
248
249
250
251
252
253
254
255
256
257
258
259
260
261
262
263
264
265
266
267
268
269
270
271
272
273
274
275
276
277


In the three cells below we create a categorical variable variable for the class label and concatenate our data to have a target matrix and an input matrix.

In [8]:
goodLabels = np.repeat("Good",len(goodArt_Transformed))
badLabels = np.repeat("Bad",len(badArt_Transformed))

In [9]:
goodArt = np.array(goodArt_Transformed)
goodLabels = np.array(goodLabels)
good = np.append(goodArt,goodLabels[:,None],axis=1)

badArt = np.array(badArt_Transformed)
badLabels = np.array(badLabels)
bad = np.append(badArt,badLabels[:,None],axis=1)

latte_pics_transformed = np.append(good,bad)

In [10]:
latte_data = np.concatenate((goodArt,badArt),axis=0)
latte_labels = np.concatenate((goodLabels,badLabels),axis=0)

We use the standard train_test_split function to generate a random test set and training set, with 80% of our images going towards training.

In [11]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(latte_data,latte_labels,test_size=0.2)

The first model we will try is a Random Forest model, with the all of the default settings. We will also use the GridSearchCV method to find optimal parameters.

In [3]:
from sklearn.ensemble import RandomForestClassifier

rForest = RandomForestClassifier(warm_start=True)
rForest.fit(X_train,y_train)

NameError: name 'X_train' is not defined

In [64]:
from sklearn.model_selection import GridSearchCV
parameters = {'n_estimators':(10,20,50,100,200),'max_depth':(2,10,20,50,100,None),'class_weight':("balanced","balanced_subsample",None)}

rForestCV = RandomForestClassifier(verbose=True)
clf = GridSearchCV(rForestCV,parameters)
clf.fit(X_train,y_train)

[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  10 out of  10 | elapsed:    0.3s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  10 out of  10 | elapsed:    0.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  10 out of  10 | elapsed:    0.3s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  10 out of  10 | elapsed:    0.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  10 out of  10 | elapsed:    0.2s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  10 out of  10 | elapsed:    0.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_j

[Parallel(n_jobs=1)]: Done  10 out of  10 | elapsed:    0.9s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  10 out of  10 | elapsed:    0.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  10 out of  10 | elapsed:    0.9s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  10 out of  10 | elapsed:    0.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  10 out of  10 | elapsed:    0.9s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  10 out of  10 | elapsed:    0.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  20 out of  20 | elapsed:    1.8s finished
[Parallel(n_jobs=1)]: Us

[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  20 out of  20 | elapsed:    2.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  20 out of  20 | elapsed:    0.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  20 out of  20 | elapsed:    2.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  20 out of  20 | elapsed:    0.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  20 out of  20 | elapsed:    2.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  20 out of  20 | elapsed:    0.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_j

[Parallel(n_jobs=1)]: Done  20 out of  20 | elapsed:    2.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  20 out of  20 | elapsed:    0.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  20 out of  20 | elapsed:    2.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  20 out of  20 | elapsed:    0.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  20 out of  20 | elapsed:    1.9s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  20 out of  20 | elapsed:    0.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  50 out of  50 | elapsed:    5.1s finished
[Parallel(n_jobs=1)]: Us

[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  50 out of  50 | elapsed:    5.1s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  50 out of  50 | elapsed:    0.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  50 out of  50 | elapsed:    5.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  50 out of  50 | elapsed:    0.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  50 out of  50 | elapsed:    5.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  50 out of  50 | elapsed:    0.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_j

[Parallel(n_jobs=1)]: Done  50 out of  50 | elapsed:    4.9s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  50 out of  50 | elapsed:    0.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  50 out of  50 | elapsed:    4.9s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  50 out of  50 | elapsed:    0.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  50 out of  50 | elapsed:    4.8s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  50 out of  50 | elapsed:    0.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done 100 out of 100 | elapsed:   10.6s finished
[Parallel(n_jobs=1)]: Us

[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done 100 out of 100 | elapsed:    2.5s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done 100 out of 100 | elapsed:    0.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done 100 out of 100 | elapsed:    2.4s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done 100 out of 100 | elapsed:    0.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done 100 out of 100 | elapsed:    2.4s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done 100 out of 100 | elapsed:    0.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_j

[Parallel(n_jobs=1)]: Done 100 out of 100 | elapsed:    9.1s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done 100 out of 100 | elapsed:    0.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done 100 out of 100 | elapsed:    9.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done 100 out of 100 | elapsed:    0.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done 100 out of 100 | elapsed:    9.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done 100 out of 100 | elapsed:    0.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done 200 out of 200 | elapsed:   18.3s finished
[Parallel(n_jobs=1)]: Us

[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done 200 out of 200 | elapsed:   20.8s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done 200 out of 200 | elapsed:    0.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done 200 out of 200 | elapsed:   20.5s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done 200 out of 200 | elapsed:    0.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done 200 out of 200 | elapsed:   20.3s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done 200 out of 200 | elapsed:    0.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_j

[Parallel(n_jobs=1)]: Done 200 out of 200 | elapsed:   20.1s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done 200 out of 200 | elapsed:    0.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done 200 out of 200 | elapsed:   20.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done 200 out of 200 | elapsed:    0.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done 200 out of 200 | elapsed:   19.5s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done 200 out of 200 | elapsed:    0.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  10 out of  10 | elapsed:    1.0s finished
[Parallel(n_jobs=1)]: Us

[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  10 out of  10 | elapsed:    1.9s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  10 out of  10 | elapsed:    0.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  10 out of  10 | elapsed:    1.8s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  10 out of  10 | elapsed:    0.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  10 out of  10 | elapsed:    1.9s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  10 out of  10 | elapsed:    0.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_j

[Parallel(n_jobs=1)]: Done  10 out of  10 | elapsed:    0.2s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  10 out of  10 | elapsed:    0.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  10 out of  10 | elapsed:    0.2s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  10 out of  10 | elapsed:    0.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  10 out of  10 | elapsed:    0.2s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  10 out of  10 | elapsed:    0.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  20 out of  20 | elapsed:    0.5s finished
[Parallel(n_jobs=1)]: Us

[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  20 out of  20 | elapsed:    1.8s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  20 out of  20 | elapsed:    0.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  20 out of  20 | elapsed:    1.7s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  20 out of  20 | elapsed:    0.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  20 out of  20 | elapsed:    1.7s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  20 out of  20 | elapsed:    0.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_j

[Parallel(n_jobs=1)]: Done  20 out of  20 | elapsed:    2.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  20 out of  20 | elapsed:    0.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  20 out of  20 | elapsed:    2.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  20 out of  20 | elapsed:    0.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  20 out of  20 | elapsed:    1.9s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  20 out of  20 | elapsed:    0.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  50 out of  50 | elapsed:    5.1s finished
[Parallel(n_jobs=1)]: Us

[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  50 out of  50 | elapsed:    5.3s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  50 out of  50 | elapsed:    0.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  50 out of  50 | elapsed:    5.5s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  50 out of  50 | elapsed:    0.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  50 out of  50 | elapsed:    5.6s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  50 out of  50 | elapsed:    0.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_j

[Parallel(n_jobs=1)]: Done  50 out of  50 | elapsed:    5.1s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  50 out of  50 | elapsed:    0.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  50 out of  50 | elapsed:    5.1s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  50 out of  50 | elapsed:    0.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  50 out of  50 | elapsed:    4.8s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  50 out of  50 | elapsed:    0.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done 100 out of 100 | elapsed:   10.0s finished
[Parallel(n_jobs=1)]: Us

[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done 100 out of 100 | elapsed:   10.3s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done 100 out of 100 | elapsed:    0.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done 100 out of 100 | elapsed:   10.3s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done 100 out of 100 | elapsed:    0.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done 100 out of 100 | elapsed:   10.9s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done 100 out of 100 | elapsed:    0.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_j

GridSearchCV(cv=None, error_score=nan,
             estimator=RandomForestClassifier(bootstrap=True, ccp_alpha=0.0,
                                              class_weight=None,
                                              criterion='gini', max_depth=None,
                                              max_features='auto',
                                              max_leaf_nodes=None,
                                              max_samples=None,
                                              min_impurity_decrease=0.0,
                                              min_impurity_split=None,
                                              min_samples_leaf=1,
                                              min_samples_split=2,
                                              min_weight_fraction_leaf=0.0,
                                              n_estimators=100, n_jobs=None,
                                              oob_score=False,
                                              ra

In [66]:
bestRF = clf.best_estimator_
bestRF.fit(X_train,y_train)
bestRF.score(X_test,y_test)

[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  20 out of  20 | elapsed:    2.8s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  20 out of  20 | elapsed:    0.0s finished


0.5444444444444444

Our best Random Forest model had a accuracy score of about 54.4%. Not much better than flipping a coin. Below we try a neural network with the standard out of the box settings and obtain an accuracy of about 58% on our test set. We use the same GridSearchCV method to find an optimal set of parameters, only changing the solver setting. This results in an accuracy of 54.8% on our test set.

In [67]:
from sklearn.neural_network import MLPClassifier

nnet = MLPClassifier()
nnet.fit(X_train,y_train)
nnet.score(X_test,y_test)

0.5814814814814815

In [68]:
parameters = {'solver':('lbfgs','sgd','adam')}
nnetCV = MLPClassifier(verbose=True)
clfNN = GridSearchCV(nnetCV,parameters)
clfNN.fit(X_train,y_train)

/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/sklearn/neural_network/_multilayer_perceptron.py:470: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
  self.n_iter_ = _check_optimize_result("lbfgs", opt_res, self.max_iter)
/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/sklearn/neural_network/_multilayer_perceptron.py:470: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
  self.n_iter_ = _check_optimize_result("lbfgs", opt_res, self.max_iter)


Iteration 1, loss = 0.91059236
Iteration 2, loss = 0.69832125
Iteration 3, loss = 0.68987983
Iteration 4, loss = 0.69289785
Iteration 5, loss = 0.68159559
Iteration 6, loss = 0.67836123
Iteration 7, loss = 0.67691695
Iteration 8, loss = 0.69702637
Iteration 9, loss = 0.66929193
Iteration 10, loss = 0.66671039
Iteration 11, loss = 0.66885485
Iteration 12, loss = 0.66011688
Iteration 13, loss = 0.67075005
Iteration 14, loss = 0.65696417
Iteration 15, loss = 0.65019912
Iteration 16, loss = 0.64834429
Iteration 17, loss = 0.64707984
Iteration 18, loss = 0.64629775
Iteration 19, loss = 0.64449479
Iteration 20, loss = 0.64165600
Iteration 21, loss = 0.63638759
Iteration 22, loss = 0.63052983
Iteration 23, loss = 0.62391928
Iteration 24, loss = 0.63337787
Iteration 25, loss = 0.63781013
Iteration 26, loss = 0.62076228
Iteration 27, loss = 0.61202245
Iteration 28, loss = 0.62058100
Iteration 29, loss = 0.62150623
Iteration 30, loss = 0.60056870
Iteration 31, loss = 0.59863379
Iteration 32, los

Iteration 155, loss = 0.29960836
Iteration 156, loss = 0.23845254
Iteration 157, loss = 0.25028166
Iteration 158, loss = 0.42121665
Iteration 159, loss = 0.23750638
Iteration 160, loss = 0.42459541
Training loss did not improve more than tol=0.000100 for 10 consecutive epochs. Stopping.
Iteration 1, loss = 0.86548253
Iteration 2, loss = 0.69051552
Iteration 3, loss = 0.68299069
Iteration 4, loss = 0.68109980
Iteration 5, loss = 0.67799564
Iteration 6, loss = 0.68214585
Iteration 7, loss = 0.67200225
Iteration 8, loss = 0.66927984
Iteration 9, loss = 0.66794254
Iteration 10, loss = 0.66364675
Iteration 11, loss = 0.66178679
Iteration 12, loss = 0.65661580
Iteration 13, loss = 0.65456472
Iteration 14, loss = 0.65510300
Iteration 15, loss = 0.64701774
Iteration 16, loss = 0.64424699
Iteration 17, loss = 0.64048291
Iteration 18, loss = 0.63503266
Iteration 19, loss = 0.63287286
Iteration 20, loss = 0.62868259
Iteration 21, loss = 0.62508105
Iteration 22, loss = 0.63411842
Iteration 23, los

Iteration 86, loss = 0.47799582
Iteration 87, loss = 0.38422405
Iteration 88, loss = 0.41859070
Iteration 89, loss = 0.42464453
Iteration 90, loss = 0.41093061
Iteration 91, loss = 0.40889106
Iteration 92, loss = 0.49804877
Iteration 93, loss = 0.43884460
Iteration 94, loss = 0.42754176
Iteration 95, loss = 0.39471759
Iteration 96, loss = 0.39488063
Iteration 97, loss = 0.48531505
Iteration 98, loss = 0.41231015
Training loss did not improve more than tol=0.000100 for 10 consecutive epochs. Stopping.
Iteration 1, loss = 0.88079522
Iteration 2, loss = 0.68973350
Iteration 3, loss = 0.68926180
Iteration 4, loss = 0.68200065
Iteration 5, loss = 0.67961151
Iteration 6, loss = 0.68484160
Iteration 7, loss = 0.67195615
Iteration 8, loss = 0.66589541
Iteration 9, loss = 0.66508860
Iteration 10, loss = 0.66642658
Iteration 11, loss = 0.67097899
Iteration 12, loss = 0.65718082
Iteration 13, loss = 0.65523245
Iteration 14, loss = 0.65779530
Iteration 15, loss = 0.64760526
Iteration 16, loss = 0.

Iteration 6, loss = 0.69440781
Iteration 7, loss = 0.69267889
Iteration 8, loss = 0.69269705
Iteration 9, loss = 0.69269437
Iteration 10, loss = 0.69269679
Iteration 11, loss = 0.69270224
Iteration 12, loss = 0.69270218
Iteration 13, loss = 0.69270278
Iteration 14, loss = 0.69270242
Iteration 15, loss = 0.69270345
Iteration 16, loss = 0.69270219
Iteration 17, loss = 0.69270197
Iteration 18, loss = 0.69270191
Training loss did not improve more than tol=0.000100 for 10 consecutive epochs. Stopping.
Iteration 1, loss = inf
Iteration 2, loss = 7.42557166
Iteration 3, loss = 4.23358824
Iteration 4, loss = 3.11020126
Iteration 5, loss = 2.10025264
Iteration 6, loss = 1.45179645
Iteration 7, loss = 1.05804961
Iteration 8, loss = 1.02696932
Iteration 9, loss = 0.93761584
Iteration 10, loss = 0.97914560
Iteration 11, loss = 1.10641735
Iteration 12, loss = 1.04222383
Iteration 13, loss = 0.76700833
Iteration 14, loss = 0.66322045
Iteration 15, loss = 0.61857418
Iteration 16, loss = 0.61101363
It

Iteration 180, loss = 0.18751043
Iteration 181, loss = 0.18172360
Iteration 182, loss = 0.54374358
Iteration 183, loss = 0.24652298
Iteration 184, loss = 0.20941921
Iteration 185, loss = 0.17486633
Iteration 186, loss = 0.20782319
Iteration 187, loss = 0.31838325
Iteration 188, loss = 0.18059235
Iteration 189, loss = 0.20452051
Training loss did not improve more than tol=0.000100 for 10 consecutive epochs. Stopping.


GridSearchCV(cv=None, error_score=nan,
             estimator=MLPClassifier(activation='relu', alpha=0.0001,
                                     batch_size='auto', beta_1=0.9,
                                     beta_2=0.999, early_stopping=False,
                                     epsilon=1e-08, hidden_layer_sizes=(100,),
                                     learning_rate='constant',
                                     learning_rate_init=0.001, max_fun=15000,
                                     max_iter=200, momentum=0.9,
                                     n_iter_no_change=10,
                                     nesterovs_momentum=True, power_t=0.5,
                                     random_state=None, shuffle=True,
                                     solver='adam', tol=0.0001,
                                     validation_fraction=0.1, verbose=True,
                                     warm_start=False),
             iid='deprecated', n_jobs=None,
             param_gri

In [69]:
bestNN = clfNN.best_estimator_
bestNN.fit(X_train,y_train)
bestNN.score(X_test,y_test)

Iteration 1, loss = 1.08143234
Iteration 2, loss = 0.69933509
Iteration 3, loss = 0.69141300
Iteration 4, loss = 0.68298134
Iteration 5, loss = 0.68286013
Iteration 6, loss = 0.68026058
Iteration 7, loss = 0.67687802
Iteration 8, loss = 0.67838157
Iteration 9, loss = 0.67197408
Iteration 10, loss = 0.67691224
Iteration 11, loss = 0.66503552
Iteration 12, loss = 0.66287071
Iteration 13, loss = 0.65516859
Iteration 14, loss = 0.65977725
Iteration 15, loss = 0.65801150
Iteration 16, loss = 0.65925074
Iteration 17, loss = 0.65341803
Iteration 18, loss = 0.64393465
Iteration 19, loss = 0.64095866
Iteration 20, loss = 0.63689401
Iteration 21, loss = 0.63288107
Iteration 22, loss = 0.63019936
Iteration 23, loss = 0.62737397
Iteration 24, loss = 0.62981183
Iteration 25, loss = 0.62264930
Iteration 26, loss = 0.61692939
Iteration 27, loss = 0.61579895
Iteration 28, loss = 0.61202572
Iteration 29, loss = 0.60698287
Iteration 30, loss = 0.60171948
Iteration 31, loss = 0.59673169
Iteration 32, los

/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/sklearn/neural_network/_multilayer_perceptron.py:568: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


0.5481481481481482

The next model we try is the KNN model and we try again to do a search to optimize the number of neighbors we use. As it turns out the GridSearchCV returns a value of 5 for K, which produces an accuracy of 57.4% on our test set.

In [71]:
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier(n_neighbors=5)
knn.fit(X_train,y_train)
knn.score(X_test,y_test)

0.5555555555555556

In [73]:
parameters = {'n_neighbors':(1,3,5,7,9)}
knnCV = KNeighborsClassifier()
clfCV = GridSearchCV(knnCV,parameters)
clfCV.fit(X_train,y_train)

GridSearchCV(cv=None, error_score=nan,
             estimator=KNeighborsClassifier(algorithm='auto', leaf_size=30,
                                            metric='minkowski',
                                            metric_params=None, n_jobs=None,
                                            n_neighbors=5, p=2,
                                            weights='uniform'),
             iid='deprecated', n_jobs=None,
             param_grid={'n_neighbors': (1, 3, 5, 7, 9)},
             pre_dispatch='2*n_jobs', refit=True, return_train_score=False,
             scoring=None, verbose=0)

In [75]:
bestKNN = clfCV.best_estimator_
bestKNN.fit(X_train,y_train)
bestKNN.score(X_test,y_test)

0.5740740740740741

We try to run the gaussian and multinomial naive-bayes models on our data and return with similar results.

In [80]:
from sklearn.naive_bayes import MultinomialNB, GaussianNB

nb = MultinomialNB()
nb.fit(X_train,y_train)
print("Multinomial Naive-Bayes")
print("Score on test: ",nb.score(X_test,y_test))
print("Score on train: ",nb.score(X_train,y_train))

gnb = GaussianNB()
gnb.fit(X_train,y_train)
print("Gaussian Naive-Bayes")
print("Score on test: ",gnb.score(X_test,y_test))
print("Score on train: ",gnb.score(X_train,y_train))

Multinomial Naive-Bayes
Score on test:  0.5037037037037037
Score on train:  0.5482374768089053
Gaussian Naive-Bayes
Score on test:  0.5444444444444444
Score on train:  0.5732838589981447


Nearly identical results are obtained with an AdaBoost model. These models take quite a long time to train so we chose 10, 50, and 100 as values for the number of estimators. As it turns out, 10 estimators produces the best accuracy, but again only 54.8%. 

In [81]:
from sklearn.ensemble import AdaBoostClassifier

ada = AdaBoostClassifier()
ada.fit(X_train,y_train)
print("Ada Boost Classifier")
print("Score on test: ",ada.score(X_test,y_test))
print("Score on train: ",ada.score(X_train,y_train))

Ada Boost Classifier
Score on test:  0.5074074074074074
Score on train:  0.87569573283859


In [84]:
ada10 = AdaBoostClassifier(n_estimators=10)
ada10.fit(X_train,y_train)
ada10.score(X_test,y_test)

0.5481481481481482

In [85]:
ada100 = AdaBoostClassifier(n_estimators=100)
ada100.fit(X_train,y_train)
ada100.score(X_test,y_test)

0.5185185185185185

Finally, we try a Decision Tree. With the standard model, we see that over model severaly overfits on our training data (high accuracy on training set and low accuracy on test set).

In [12]:
from sklearn.tree import DecisionTreeClassifier

dt = DecisionTreeClassifier()
dt.fit(X_train,y_train)
print("Decision Tree Classifier (Default)")
print("Score on test",dt.score(X_test,y_test))
print("Score on train",dt.score(X_train,y_train))

Decision Tree Classifier (Default)
Score on test 0.5296296296296297
Score on train 0.9953617810760668


In [16]:
dt2 = DecisionTreeClassifier(max_depth=5)
dt2.fit(X_train,y_train)
print("Decision Tree Classifier (Default)")
print("Score on test",dt2.score(X_test,y_test))
print("Score on train",dt2.score(X_train,y_train))

Decision Tree Classifier (Default)
Score on test 0.5444444444444444
Score on train 0.7430426716141002
